bindings/rethinkdb/statechange/statechange.go

```diff
@@ -0,0 +1,187 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package statechange
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+	"time"
+
+	r "github.com/dancannon/gorethink"
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/pkg/errors"
+)
+
+// Binding represents RethinkDB change change state input binding which fires handler with
+// both the previous and current state store content each time there is a change.
+type Binding struct {
+	logger  logger.Logger
+	session *r.Session
+	config  StateConfig
+	stopCh  chan bool
+}
+
+// StateConfig is the binding config
+type StateConfig struct {
+	r.ConnectOpts
+	Table string `json:"table"`
+}
+
+var _ = bindings.InputBinding(&Binding{})
+
+// NewRethinkDBStateChangeBinding returns a new RethinkDB actor event input binding
+func NewRethinkDBStateChangeBinding(logger logger.Logger) *Binding {
+	return &Binding{
+		logger: logger,
+		stopCh: make(chan bool),
+	}
+}
+
+// Init initializes the RethinkDB binding
+func (b *Binding) Init(metadata bindings.Metadata) error {
+	cfg, err := metadataToConfig(metadata.Properties, b.logger)
+	if err != nil {
+		return errors.Wrap(err, "unable to parse metadata properties")
+	}
+	b.config = cfg
+
+	ses, err := r.Connect(b.config.ConnectOpts)
+	if err != nil {
+		return errors.Wrap(err, "error connecting to the database")
+	}
+	b.session = ses
+
+	return nil
+}
+
+// Read triggers the RethinkDB scheduler
+func (b *Binding) Read(handler func(*bindings.ReadResponse) error) error {
+	b.logger.Infof("subscribing to state changes in %s.%s...", b.config.Database, b.config.Table)
+	cursor, err := r.DB(b.config.Database).Table(b.config.Table).Changes(r.ChangesOpts{
+		IncludeTypes: true,
+	}).Run(b.session)
+	if err != nil {
+		errors.Wrapf(err, "error connecting to table %s", b.config.Table)
+	}
+
+	go func() {
+		for {
+			var change interface{}
+			ok := cursor.Next(&change)
+			if !ok {
+				b.logger.Errorf("error detecting change: %v", cursor.Err())
+				break
+			}
+
+			data, err := json.Marshal(change)
+			if err != nil {
+				b.logger.Errorf("error marshalling change handler: %v", err)
+			}
+			b.logger.Debugf("event: %s", string(data))
+
+			resp := &bindings.ReadResponse{
+				Data: data,
+				Metadata: map[string]string{
+					"store-address":  b.config.Address,
+					"store-database": b.config.Database,
+					"store-table":    b.config.Table,
+				},
+			}
+
+			if err := handler(resp); err != nil {
+				b.logger.Errorf("error invoking change handler: %v", err)
+				continue
+			}
+		}
+	}()
+
+	done := <-b.stopCh
+	b.logger.Errorf("done: %b", done)
+	defer cursor.Close()
+	return nil
+}
+
+func metadataToConfig(cfg map[string]string, logger logger.Logger) (StateConfig, error) {
```

Comment:
Is there any good way to not duplicate the same code ?

Reply:
These aren't exactly the same.... Also, I'm not sure we want to encourage direct inter-component dependencies. Increased blast radius on subsequent changes and potential resulting (unintended) side effects, probably not worth to prevent dup of 10-20 lines of generic code. 

Maybe the answer here is some global "helper" package in the "components-contrib" itself that can be used for common/duplicative tasks like parsing metadata that components can use if they want to?